docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,331 @@
+10 shares. One year later, 1 share is worth 120 tokens, so if I transfer
+1,000 tokens, the program converts and only moves 8.33 shares. The same concept
+applies to mints and burns.
+
+Following this model, we never have to worry about compounding interest. We
+simply need to call the right program to perform instructions.
+
+## Required Changes
+
+A new interest-bearing token program is only one component of the overall solution,
+since i-tokens need to be easily integrated in on-chain programs, web apps,
+and wallets.
+
+### Token Program Conformance
+
+Since our solution entails the creation of a new token program, we're opening the
+floodgates for other app developers to do the same. To ensure safe and easy integration
+of new token programs, we need a set of conformance tests based on the current
+SPL token test suite.
+
+A new token program must include all of the same instructions as SPL token,
+but the "unchecked" variants should return errors. Concretely, this means the
+program must implement:
+
+* `InitializeMint`
+* `InitializeHolding` (previously `InitializeAccount`)
+* `Revoke`
+* `SetAuthority`
+* `CloseHolding` (previously `CloseAccount`)
+* `FreezeHolding` (previously `FreezeAccount`)
+* `ThawHolding` (previously `ThawAccount`)
+* `TransferChecked`
+* `ApproveChecked`
+* `MintToChecked`
+* `BurnChecked`
+* `InitializeHolding2` (previously `InitializeAccount2`)
+
+And the program should throw errors for:
+
+* `Transfer`
+* `Approve`
+* `MintTo`
+* `Burn`
+* `InitializeMultisig` (use a more general multisig)
+
+New instructions required:
+
+* `CreateHolding` (TODO name): only performs `Allocate` and `Assign` to self,
+useful for creating a holding when you don't know the program that you're
+interacting with. See the Associated Token Account section for how to use this.
+
+There are also new read-only instructions, which write data to a provided account
+buffer:
+
+* `GetHoldingBalance`: balance of the holding
+* `GetHoldingOwner`: owner's public key
+* `GetMintSupply`: mint supply
+* `GetMintAuthority`: mint authority's public key
+* `GetMintDecimals`: mint decimals
+
+See the Runtime section for more information on how these are used.
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+The SPL token library provides wrappers compatible with any conforming program.
+For the base SPL token program, the wrapper simply deserializes the account data
+and returns the relevant data. For other programs, it calls the appropriate
+read-only instruction using an ephemeral account input.
+
+The structure for new holding and mint accounts must follow the layout of the
+existing SPL token accounts to ensure compatibility in RPC when fetching pre and
+post token balances.
+
+For a holding, the first 165 bytes must contain the same information
+as a normal SPL token holding. The following byte must be the type of
```

Comment:
is this spoofing-attack safe? What if a malicious token program populates its token holding account with the mint byte range being initialized with one of established spl-tokens?

Reply:
I want to be sure that I understand this -- you mean a spoofing attack where a malicious token program intentionally mixes up tokens and mints?  To me, that would mean the malicious token program doesn't work at all, not impacting other tokens.  How would it relate to the standard spl token in that case?

Reply:
this is rather minor issue.

But, malicious program and its data accounts can minic as if it's a (hypothetical) well established tokens like USDZ living in the tokenkeg land. That's because malicious token program has completely control the bits.

Then, further assume that victim program is hard-coding mint address of USDZ but they're not checking account.owner/program_id as being tokenkeg. This  could trick the problem.

being `minor` is that because `TransferChecked` would fail anyway. But (unusual?) custody model might cause problem (by set_authority) because that might not write the faked account data unlike TransferChecked.

So, I think we should strongly recommend to check program_id everywhere and actually add these checks across the spl-token / i-token codebase.

So, they can even bypass the base spl-token deserialize check by adding 0 extra data intentionally. (So, at least 1 more byte worth nice security protection)

Also, `SerializeToSPL` could return whatever faked binaries.

Reply:
I see what you mean now.  That's definitely a consideration, and the best solution seems to force people to always check the program id of the account.  The new wrapper library will accept an `AccountInfo` for the token program, so we'll do the check there, which should protect most users.  There's still a risk of bad token programs being out there, which is why the registry will become an important component.

Reply:
Agreed with @ryoqun that the possibility of mixing up mints and holding accounts is undesirable. I think the design should make it hard for token program designers to mess this up. Reserving the first byte for determining which account is being deserialized seems like the best approach (especially when custom token programs define other account types). Judging by account size was never meant to be a general approach.